integration_tests/src/main/python/cpu_bridge_test.py

```diff
@@ -0,0 +1,219 @@
+    assert_gpu_and_cpu_are_equal_collect(test_func, conf=conf)
+
+# This is borrowed partly from join_test.py
+bloom_filter_confs = {
+    "spark.sql.autoBroadcastJoinThreshold": 1,
+    "spark.sql.optimizer.runtime.bloomFilter.applicationSideScanSizeThreshold": 1,
+    "spark.sql.optimizer.runtime.bloomFilter.creationSideThreshold": "100GB",
+    "spark.sql.optimizer.runtime.bloomFilter.enabled": True
+}
+
+def check_bloom_filter_join(confs, is_multi_column):
+    def do_join(spark):
+        if is_multi_column:
+            left = spark.range(100000).withColumn("second_id", col("id") % 5)
+            right = spark.range(10).withColumn("id2", col("id").cast("string")).withColumn("second_id", col("id") % 5)
+            return right.filter("cast(id2 as bigint) % 3 = 0").join(left, (left.id == right.id) & (left.second_id == right.second_id), "inner")
+        else:
+            left = spark.range(100000)
+            right = spark.range(10).withColumn("id2", col("id").cast("string"))
+            return right.filter("cast(id2 as bigint) % 3 = 0").join(left, left.id == right.id, "inner")
+    bridge_conf = create_cpu_bridge_fallback_conf([])
+    partial_conf = copy_and_update(bridge_conf, confs)
+    all_confs = copy_and_update(bloom_filter_confs, partial_conf)
+    assert_gpu_and_cpu_are_equal_collect(do_join, conf=all_confs)
+
+@allow_non_gpu("ShuffleExchangeExec")
+@ignore_order(local=True)
+@pytest.mark.parametrize("is_multi_column", [False, True], ids=["SINGLE_COLUMN", "MULTI_COLUMN"])
+@pytest.mark.skipif(is_databricks_runtime(), reason="https://github.com/NVIDIA/spark-rapids/issues/8921")
+@pytest.mark.skipif(is_before_spark_330(), reason="Bloom filter joins added in Spark 3.3.0")
+def test_bloom_filter_join_cpu_probe(is_multi_column):
+    conf = {"spark.rapids.sql.expression.BloomFilterMightContain": "false"}
+    check_bloom_filter_join(confs=conf, is_multi_column=is_multi_column)
+
+# ==============================================================================
+# NEGATIVE TEST CASES - Verify expressions that should NOT use CPU bridge
+# These should cause full CPU fallback instead of using the bridge
+# ==============================================================================
+
+@allow_non_gpu('ProjectExec')
+def test_cpu_bridge_rand_disabled_fallback():
+    """Test that when rand() is disabled via config, ProjectExec falls back to CPU entirely"""
+    def test_func(spark):
+        df = gen_df(spark, [('a', int_gen), ('b', int_gen)], length=1000)
+        # rand(42) with seed for deterministic results - disabled via config should cause full CPU fallback
+        return df.selectExpr("a", "b", "a + b as sum", "rand(42) as random_val")
+    
+    # Enable CPU bridge but disable rand() - should cause full ProjectExec fallback
+    conf = create_cpu_bridge_fallback_conf(['Rand'])
+    
+    # Verify that ProjectExec falls back to CPU (doesn't use GPU or bridge)
+    assert_gpu_fallback_collect(test_func, 'ProjectExec', conf=conf)
```

Comment:
These tests actually verify that we did the fallback to cpu, in addition to the results being equal. The above several tests implicitly expect that we use the bridge and assert the results are equal, but should we actually explicitly assert that we did not fallback/did use the bridge in those cases?